### PR TITLE
remove_qubit method for faster statevec simulation

### PR DESCRIPTION
**Pull request by nabe98**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
- format added code and tests by `black -l 120 <filename>`
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
 
Then, please fill in below:

**Context (if applicable):**
`ptrace` method was implemented by using density matrix. But for separable states (states that measured), using DM is unnecessary. 

**Description of the change:**

 - Add `truncate_one_qubit` method to `Statevec`. 
 
directly truncate the specified qubit.

**Related issue:**
#73 

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

**Review by masa10-f — COMMENTED** [review]
Could you add more test cases?
At least include the following measurement commands
1. measure into |0>
2. measure into |1>
3. non-Pauli measurement

**Comment by masa10-f** [comment]
Could you change the `measure` method in the `StatevectorBackend` to trace out measured qubits individually?

**Comment by masa10-f** [comment]
Thanks for your commitment!
Your code looks great, but I would like to mention one minor point: the state vector backend should not depend on the pattern or circuit simulator. Therefore, it's preferable to avoid them in the unit tests. So, I have removed the random circuit check.

**Review by masa10-f — APPROVED** [review]

**Comment by masa10-f** [comment]
@nabe98  Before merging, can you write changes to CHANGELOG.md?

File: graphix/sim/statevec.py
Comment (shinich1) [inline graphix/sim/statevec.py:2]: please remove extra line here.

File: graphix/sim/statevec.py
Comment (shinich1) [inline graphix/sim/statevec.py:255]: let me fix up this text later, reads a bit strange?

File: graphix/sim/statevec.py
Comment (shinich1) [inline graphix/sim/statevec.py:388]: why not add as a method (eg `_get_norm` if you don't want to explicitly expose it

File: graphix/sim/statevec.py
Comment (shinich1) [inline graphix/sim/statevec.py:271]: is `truncate` a good word here? let me have a think for a bit

Comment (shinich1) [reply graphix/sim/statevec.py:271]: truncate sounds like removing the last qubit?

File: tests/test_statevec_backend.py
Comment (shinich1) [inline tests/test_statevec_backend.py:10]: is this supposed to be complete testing of the backend, or is this only for `truncate` method?

**Comment by shinich1** [comment]
@nabe98 please let me know once you updated CHANGELOG.md. Please do not merge yet. 
Also remember we always `prefer squash and merge` to keep the commit history concise on master branch.

File: graphix/sim/statevec.py
Comment (masa10-f) [reply graphix/sim/statevec.py:2]: I followed `isort` package here. Why don't we obey PEP8?

File: graphix/sim/statevec.py
Comment (masa10-f) [reply graphix/sim/statevec.py:197]: Because we can generate Statevec object simply with `Statevec(n)`. I think this is a standard format.

File: graphix/sim/statevec.py
Comment (masa10-f) [reply graphix/sim/statevec.py:271]: Hmm... `discard` or `remove` ?

File: graphix/sim/statevec.py
Comment (shinich1) [reply graphix/sim/statevec.py:2]: sure if there's good reason!

File: tests/test_statevec_backend.py
Comment (masa10-f) [reply tests/test_statevec_backend.py:10]: supposed to be complete testing of the statevector backend. In the future, when we add another method into statevec, we will make another tests on this file.

File: graphix/sim/statevec.py
Comment (shinich1) [inline graphix/sim/statevec.py:276]: @nabe98 does this sound correct? could you check? 
also could you make sure that docs for this function generates as intended (you'll need to `make html` in `docs/` and check for simulator API doc). I have not checked on my side, I would appreciate if you could debug and check that docs gen work!

File: graphix/sim/statevec.py
Comment (nabe98) [reply graphix/sim/statevec.py:403]: `_get_statevec_norm` sounds preferable to me!

Wildcard import is not recommended (https://peps.python.org/pep-0008/#imports) so we should avoid that.

File: graphix/sim/statevec.py
Comment (nabe98) [reply graphix/sim/statevec.py]: @shinich1 
I have updated the docs slightliy and see that the intended HTML is generated correctly.

**Comment by nabe98** [comment]
@shinich1 also updated `CHANGELOG.md`. Pleas check them!

File: graphix/sim/statevec.py
Comment (shinich1) [reply graphix/sim/statevec.py]: thank you!

**Comment by masa10-f** [comment]
I've conducted benchmarks for the new method.

[Environment]
CPU: intel i7-12700K
RAM: 64GB (DDR4-3200)
OS: Ubuntu 20.04 (WSL)

The following graph shows a performance comparison between `reduce_qubit` and `ptrace` when simulating n (width) x n (depth) random square circuits.
svbackend_comparison_square_circuit

I also benchmarked shallow (depth=1) random circuits, only changing the number of qubits.
svbackend_comparison_shallow

**Comment by masa10-f** [comment]
I used the following random circuit generator

```
def random_circuit(nqubit, depth, seed=None):
    r"""Generate a random circuit.

    This function generates a random circuit with nqubit qubits and depth layers.
    The circuit is generated by CNOT gates and RZ gates.

    Parameters
    ----------
    nqubit : int
        number of qubits
    depth : int
        number of layers
    seed : int
        random seed

    Returns
    -------
    circuit : graphix.transpiler.Circuit object
        generated circuit
    """
    if seed is not None:
        random.seed(seed)
    qubit_index = [i for i in range(nqubit)]
    circuit = Circuit(nqubit)
    for _ in range(depth):
        random.shuffle(qubit_index)
        for j in range(len(qubit_index) // 2):
            circuit.cnot(qubit_index[2 * j], qubit_index[2 * j + 1])
        for j in range(len(qubit_index)):
            circuit.rz(qubit_index[j], 2 * np.pi * random.random())
    return circuit
```

**Comment by masa10-f** [comment]
I pasted the benchmark code below. I didn't perform Pauli measurements because `Pattern.minimize_space()` for a Pauli-measured pattern sometimes doesn't work well.

```
# %%
from graphix.generator import random_circuit
import matplotlib.pyplot as plt

from time import perf_counter

# %%
test_cases = [(i, 1) for i in range(2, 30, 2)]

score_time = []
circuit_time = []

# %%
for case in test_cases:
    nqubit, depth = case
    circuit = random_circuit(nqubit, depth)
    pattern = circuit.transpile()
    pattern.standardize()
    pattern.minimize_space()
    print(f"max space for nqubit={nqubit} circuit is ", pattern.max_space())
    start = perf_counter()
    pattern.simulate_pattern()
    end = perf_counter()
    print(f"nqubit: {nqubit}, depth: {depth}, time: {end - start}")
    score_time.append(end - start)
    start = perf_counter()
    circuit.simulate_statevector()
    end = perf_counter()
    circuit_time.append(end - start)

# %%
# plot the score_time
plt.scatter([case[0] for case in test_cases], score_time, label="score_time")
plt.scatter([case[0] for case in test_cases],
            circuit_time, label="circuit_time")
plt.xlabel("nqubit")
plt.ylabel("time (s)")
# plt.xscale("log")
plt.yscale("log")
plt.title("Time to simulate square random circuits")
plt.legend()
plt.show()

# %%
# write to file
with open("score_time.txt", "w") as f:
    for item in score_time:
        f.write(f"{item}\n")

# %%
# write circuit_time to file
with open("circuit_time.txt", "w") as f:
    for item in circuit_time:
        f.write(f"{item}\n")
```

**Review by shinich1 — APPROVED** [review]

**Comment by shinich1** [comment]
@nabe98 can you squash and merge?

**Pull request merged**
